rasa/src/Rasa/Internal/Action.hs

```diff
@@ -131,14 +203,97 @@ bootstrapAction action = do
+                buffers.at bIndex ?= newBuffer
+                return res
+
+    (AddListener listenerF withListenerId) -> do
+      n <- nextListenerId <<+= 1
+      let (listener, listenerId, eventType) = mkListener n listenerF
+      listeners %= M.insertWith mappend eventType [listener]
+      actionInterpreter $ withListenerId listenerId
+        where
+          mkListener :: forall event r. Typeable event => Int -> (event -> Action r) -> (Listener, ListenerId, TypeRep)
+          mkListener n listenerFunc =
+            let list = Listener listId (void . listenerFunc)
+                listId = ListenerId n (typeRep (Proxy :: Proxy event))
+                prox = typeRep (Proxy :: Proxy event)
+             in (list, listId, prox)
+
+    (RemoveListener listenerId@(ListenerId _ eventType) next) -> do
+      listeners.at eventType._Just %= filter (notMatch listenerId)
+      actionInterpreter next
+        where
+          notMatch idA (Listener idB _) = idA /= idB
+
+    (DispatchEvent evt next) -> do
+      listeners' <- use listeners
+      let (Action action) = traverse_ ($ evt) (matchingListeners listeners')
+      actionInterpreter (action >> next)
+
+    (DispatchActionAsync asyncActionIO next) -> do
+      asyncQueue <- use actionQueue
+      let effect = (liftIO asyncActionIO >>= yield) >-> toOutput asyncQueue
+      liftIO $ forkIO $ runEffect effect >> performGC
+      actionInterpreter next
+
+    (AsyncActionProvider dispatcherToIO next) -> do
+      asyncQueue <- use actionQueue
+      let dispatcher action =
+            let effect = yield action >-> toOutput asyncQueue
+             in void . forkIO $ runEffect effect >> performGC
+      liftIO . forkIO $ dispatcherToIO dispatcher
+      actionInterpreter next
+
+    (AddBuffer toNext) -> do
+      bufId <- nextBufId <+= 1
+      buffers.at bufId ?= def
+      actionInterpreter . toNext $ BufRef bufId
+
+    (GetBufRefs toNext) ->
+      use (buffers.to IM.keys) >>= actionInterpreter . toNext . fmap BufRef
+
+    (GetExt toNext) ->
+      use ext >>= actionInterpreter . toNext
+
+    (SetExt new next) -> do
+      ext .= new
+      actionInterpreter next
+
+    (GetEditor toNext) ->
+      use ed >>= actionInterpreter . toNext
+
+    (GetBuffer (BufRef bufInd) toNext) ->
+      use (buffers.at bufInd) >>= actionInterpreter . toNext
+
+    (Exit next) -> do
+      exiting .= True
+      actionInterpreter next
+
+    (ShouldExit toNext) -> do
+      ex <- use exiting
+      actionInterpreter (toNext ex)
+
+actionInterpreter (Pure res) = return res
+
+-- | This extracts all event listeners from a map of listeners which match the type of the provided event.
+matchingListeners :: forall a. Typeable a => Listeners -> [a -> Action ()]
+matchingListeners listeners' = getListener <$> (listeners'^.at (typeRep (Proxy :: Proxy a))._Just)
 
-actionInterpreter actionState (Pure res) = return (res, actionState)
+getListener :: forall a. Listener -> (a -> Action ())
```

Comment:
Good catch; you're referring to the existential quantifier? Apparently it can just be removed haha. So can most of the quantifiers in this file (except for the ones dealing with `Proxy`s.